segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -495,54 +533,63 @@ public void initialize() throws Exception {
-                                                log.info("SystemJournal[{}] Done finding snapshots. Snapshot found and parsed. {}", containerId, persisted);
-                                                return systemSnapshot;
-                                            } catch (Exception e) {
-                                                val ex = Exceptions.unwrap(e);
-                                                if (ex instanceof EOFException) {
-                                                    log.warn("SystemJournal[{}] Incomplete snapshot found, skipping {}.", containerId, persisted, e);
-                                                } else if (ex instanceof ChunkNotFoundException) {
-                                                    log.warn("SystemJournal[{}] Missing snapshot, skipping {}.", containerId, persisted, e);
-                                                } else {
-                                                    log.error("SystemJournal[{}] Error with snapshot, skipping {}.", containerId, persisted, e);
-                                                    throw new CompletionException(e);
-                                                }
-                                            }
-                                            return null;
-                                        }, executor),
-                                executor);
+                        return checkSnapshotExists(snapshotFileName)
+                                // Step 3: Read contents.
+                                .thenComposeAsync(v -> getContents(snapshotFileName), executor)
+                                // Step 4: Deserialize and return.
+                                .thenApplyAsync(contents -> readSnapshotRecord(snapshotInfo, contents), executor);
                     } else {
                         return CompletableFuture.completedFuture(null);
                     }
                 }, executor);
     }
 
+    /**
+     * Check whether snapshot file exists.
+     */
+    private CompletableFuture<Void> checkSnapshotExists(String snapshotFileName) {
+        if (getConfig().isSelfCheckEnabled()) {
+            return chunkStorage.exists(snapshotFileName)
+                    .thenAcceptAsync(exists -> Preconditions.checkState(exists, "File pointed by SnapshotInfo must exist"), executor);
+        } else {
+            return CompletableFuture.completedFuture(null);
+        }
+    }
+
+    /**
+     * Read {@link SystemSnapshotRecord} from given byte array.
+     */
+    private SystemSnapshotRecord readSnapshotRecord(SnapshotInfo snapshotInfo, byte[] snapshotContents) {
+        try {
+            val systemSnapshot = SYSTEM_SNAPSHOT_SERIALIZER.deserialize(snapshotContents);
+            log.info("SystemJournal[{}] Done finding snapshots. Snapshot found and parsed. {}", containerId, snapshotInfo);
```

Comment:
Do we need an `info` for each individual snapshot? Let's make this `debug` and just keep the overall summary at `info`. I don't want to pollute the recovery logs with unnecessary INFOs if not needed.

Reply:
This should happen only once during bootstrap.
Knowing actually what snapshot was read during boot is important info.

Reply:
Fixed. Added info log somewhere else